File: backend/src/main/java/com/hgw/gestionale/customer/controller/CustomerController.java
package com.hgw.gestionale.customer.controller;

import com.hgw.gestionale.customer.dto.CreateCustomerRequest;
import com.hgw.gestionale.customer.dto.CustomerResponse;
import com.hgw.gestionale.customer.dto.UpdateCustomerRequest;
import com.hgw.gestionale.customer.service.CustomerService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping(value = "/api/customers", produces = MediaType.APPLICATION_JSON_VALUE)
@RequiredArgsConstructor
@Slf4j
public class CustomerController {
    private final CustomerService customerService;

    @PostMapping
    public CustomerResponse createCustomer(@Valid @RequestBody CreateCustomerRequest request) {
        log.info("CustomerController.create creazione cliente");
        return customerService.create(request);
    }

    @GetMapping
    public List<CustomerResponse> getAllCustomers() {
        log.info("CustomerController.getAll richiesta lista clienti");
        return customerService.getAll();
    }

    @GetMapping("/search")
    public List<CustomerResponse> searchCustomers(
            @RequestParam(required = false) String nome,
            @RequestParam(required = false) String cognome,
            @RequestParam(required = false) String telefono,
            @RequestParam(required = false) String email
    ) {
        log.info("CustomerController.search nome={} cognome={} telefono={} email={}", nome, cognome, telefono, email);
        return customerService.search(nome, cognome, telefono, email);
    }

    

    @GetMapping("/{id}")
    public CustomerResponse getCustomerById(@PathVariable Long id) {
        log.info("CustomerController.getById richiesta cliente id={}", id);
        return customerService.getById(id);
    }

    @PutMapping("/{id}")
    public CustomerResponse updateCustomer(
            @PathVariable Long id,
            @Valid @RequestBody UpdateCustomerRequest request
    ) {
        log.info("CustomerController.update aggiornamento cliente id={}", id);
        return customerService.update(id, request);
    }

    @DeleteMapping("/{id}")
    public void deleteCustomer(@PathVariable Long id) {
        log.info("CustomerController.delete eliminazione cliente id={}", id);
        customerService.delete(id);
    }
}

File: backend-node/src/controllers/customerController.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  OperationId,
  Path,
  Post,
  Put,
  Query,
  Route,
  Tags,
} from "tsoa";
import {
  CustomerResponse,
  CreateCustomerRequest,
  UpdateCustomerRequest,
} from "../dto/customer.dto";

@Route("customers")
@Tags("Customers")
export class CustomerController extends Controller {
  @Post("/")
  @OperationId("createCustomer")
  public async createCustomer(@Body() body: CreateCustomerRequest): Promise<CustomerResponse> {
    throw new Error("Not implemented");
  }

  @Get("/")
  @OperationId("getAllCustomers")
  public async getAllCustomers(): Promise<CustomerResponse[]> {
    throw new Error("Not implemented");
  }

  @Get("/search")
  @OperationId("searchCustomers")
  public async searchCustomers(
    @Query() nome?: string,
    @Query() cognome?: string,
    @Query() telefono?: string,
    @Query() email?: string
  ): Promise<CustomerResponse[]> {
    throw new Error("Not implemented");
  }

  @Get("/{id}")
  @OperationId("getCustomerById")
  public async getCustomerById(@Path() id: number): Promise<CustomerResponse> {
    throw new Error("Not implemented");
  }

  @Put("/{id}")
  @OperationId("updateCustomer")
  public async updateCustomer(
    @Path() id: number,
    @Body() body: UpdateCustomerRequest
  ): Promise<CustomerResponse> {
    throw new Error("Not implemented");
  }

  @Delete("/{id}")
  @OperationId("deleteCustomer")
  public async deleteCustomer(@Path() id: number): Promise<void> {
    throw new Error("Not implemented");
  }
}
